File: week5/python/2_primitive_calculator.py
def primitive_calculator(number):
    min_steps = [None]*(number+1)
    predecessor = [None]*(number+1)

    min_steps[1] = 0

    for i in range(2, number+1):
        min_steps[i] = min_steps[i-1]+1
        predecessor[i] = i-1

        if(i % 3 == 0):
            if(min_steps[i//3] < min_steps[i]):
                min_steps[i] = min_steps[i//3] + 1
                predecessor[i] = i//3
        if(i % 2 == 0):
            if(min_steps[i//2] < min_steps[i]):
                min_steps[i] = min_steps[i//2] + 1
                predecessor[i] = i//2

    print(min_steps[number])
    i = number
    sequence = []
    while(i != 1):
        sequence.append(i)
        i = predecessor[i]

    sequence.append(1)
    print(*sequence[::-1])


number = int(input())
primitive_calculator(number)

File: week5/c++/2_primitive_calculator.cpp
#include <iostream>
#include <vector>
#include <list>
#include <algorithm>

using namespace std;

void primitive_calculator(int number)
{
    vector<int> counts(number + 1);
    vector<int> predecessor(number + 1);

    for (int i = 2; i <= number; i++)
    {
        counts[i] = counts[i - 1] + 1;
        predecessor[i] = i - 1;
        if (i % 3 == 0)
        {
            if (counts[i / 3] < counts[i])
            {
                counts[i] = counts[i / 3] + 1;
                predecessor[i] = i / 3;
            }
        }
        if (i % 2 == 0)
        {
            if (counts[i / 2] < counts[i])
            {
                counts[i] = counts[i / 2] + 1;
                predecessor[i] = i / 2;
            }
        }
    }
    cout << counts[number] << endl;

    int i = number;

    vector<int> v;

    while (i != 1)
    {
        v.push_back(i);
        i = predecessor[i];
    };

    reverse(v.begin(), v.begin() + v.size());

    cout << 1 << " ";

    for (auto i : v)
    {
        cout << i << " ";
    }
    cout << endl;

    /*
       list<int> sequence;
       for(int i = number;i!=0;i=predecessor[i]){
           sequence.push_front(i);
           */
}

int main()
{
    int number;
    cin >> number;
    primitive_calculator(number);
    return 0;
}